### ISSUE TITLE
Question: Is there a way to regenerate the ExifTool.jpg test image with a custom image?

### ISSUE THREAD

**Issue by gsuberland** [exiftool/exiftool#194]
I'm doing some testing related to image metadata stripping in various software. I had initially generated my own test JPEG with around 140 EXIF tags, but then I stumbled across ExifTool.jpg in the exiftool test images directory, which seems to include almost every standard and nonstandard tag and edge-case. The only problem is that the image itself is just a small white square, which tends to trigger PNG auto-conversion in software that has that feature.

I tried `exiftool -tagsFromFile ExifTool.jpg target.jpg` and `exiftool -m -tagsFromFile ExifTool.jpg "-all:all>all:all" target.jpg` but only a fraction of the metadata from the test image is actually copied over.

Is there a way to regenerate that test file but with custom image content?

**Comment by StarGeekSpaceNerd** [comment]
See [FAQ #9](https://exiftool.org/faq.html#Q9).  

tl;dr, use `exiftool -TagsFromFile source.jpg -All:All target.jpg`

The [Test File Args file](https://exiftool.org/forum/index.php?topic=9212.0) might also be of interest to you.  It allows you to create a test file with over 1,400 tags in it.  They are different tags than in the exiftool.jpg file, but some of those are pretty rare, IMO.

**Comment by gsuberland** [comment]
I tried that command; it essentially has the same results as the one I mentioned above. Only around 160 tags get copied across instead of the nearly 700 in the original test file.

Your args file does seem to create a lot of tags though, which is useful, thanks. The only downside for my use-case is that a large portion of the tags are XMP-based, and many metadata stripping algorithms just wholesale remove the XMP block. One of the things that drew me to the ExifTool.jpg test file is that it has a very diverse set of metadata types embedded in it.

I used `-args` on ExifTool.jpg and then merged that with your args file, which did successfully introduce a few extras, but the majority of the new arguments resulted in "Sorry, [...] doesn't exist or isn't writable" or "Tag '[...]' is not defined". I'm guessing they'd need to be block-copied from the original, but the `-tagsFromFile` argument doesn't seem to be doing that. Any ideas on what to try?

**Comment by StarGeekSpaceNerd** [comment]
> I tried that command; it essentially has the same results as the one I mentioned above.

Ah, yes, my apologies for not reading carefully enough.

Looking closer, many of those tags really are so obscure that you're extremely unlikely to ever come across them.  The best I can offer is to copy various groups as blocks with    
`exiftool -TagsFromFile source.jpg -EXIF -JFIF -icc_profile -XMP -IPTC -makernotes -trailer target.jpg`

Phil will have to comment further.

**Comment by boardhead** [comment]
I wrote a swap_image script that does the reverse: It inserts a new image in an existing file, maintaining most of the metadata:  https://exiftool.org/forum/index.php?msg=8031

**Comment by gsuberland** [comment]
I've made a bit more progress but this is still a little confusing.

First I ran `exiftool -listg1` to get a list of groups, removed the ones ending in `#`, and created this monstrosity of a command:

```
exiftool -m -TagsFromFile /path/to/ExifTool.jpg -AC3 -AFCP -AIFF -APE -ASF -AVI1 -Adobe -AdobeCM -AdobeDNG -Apple -Audible -CIFF -CameraIFD -Canon -CanonCustom -CanonRaw -CanonVRD -Casio -Composite -DICOM -DJI -DNG -DV -DjVu -DjVu-Meta -Ducky -EPPIM -EXE -EXIF -ExifIFD -ExifTool -FITS -FLAC -FLIR -File -Flash -FlashPix -Font -FotoStation -FujiFilm -FujiIFD -GE -GIF -GIMP -GPS -GeoTiff -GlobParamIFD -GoPro -GraphConv -H264 -HP -HTC -HTML -HTML-dc -HTML-ncc -HTML-office -HTML-prod -HTML-vw96 -HTTP-equiv -ICC-chrm -ICC-clrt -ICC-header -ICC-meas -ICC-meta -ICC-view -ICC_Profile -ID3 -ID3v1 -ID3v1_Enh -ID3v2_2 -ID3v2_3 -ID3v2_4 -IFD0 -IFD1 -IPTC -ISO -ITC -Insta360 -InteropIFD -ItemList -JFIF -JFXX -JPEG -JPEG-HDR -JSON -JVC -Jpeg2000 -KDC_IFD -Keys -Kodak -KodakBordersIFD -KodakEffectsIFD -KodakIFD -KyoceraRaw -LNK -Leaf -LeafSubIFD -Leica -Lytro -M2TS -MAC -MIE-Audio -MIE-Camera -MIE-Canon -MIE-Doc -MIE-Extender -MIE-Flash -MIE-GPS -MIE-Geo -MIE-Image -MIE-Lens -MIE-Main -MIE-MakerNotes -MIE-Meta -MIE-Orient -MIE-Preview -MIE-Thumbnail -MIE-UTM -MIE-Unknown -MIE-Video -MIFF -MNG -MOBI -MOI -MPC -MPEG -MPF0 -MPImage -MS-DOC -MXF -MacOS -MakerNotes -MakerUnknown -Matroska -MediaJukebox -Meta -MetaIFD -Microsoft -Minolta -MinoltaRaw -Motorola -NITF -Nikon -NikonCapture -NikonCustom -NikonScan -Nintendo -Ocad -Ogg -Olympus -OpenEXR -Opus -PDF -PICT -PNG -PNG-pHYs -PSP -Palm -Panasonic -PanasonicRaw -Parrot -Pentax -PhaseOne -PhotoCD -PhotoMechanic -Photoshop -PictureInfo -PostScript -PreviewIFD -PrintIM -ProfileIFD -Qualcomm -QuickTime -RAF -RAF2 -RIFF -RMETA -RSRC -RTF -Radiance -Rawzor -Real -Real-CONT -Real-MDPR -Real-PROP -Real-RA3 -Real-RA4 -Real-RA5 -Real-RJMD -Reconyx -Red -Ricoh -SPIFF -SR2 -SR2DataIFD -SR2SubIFD -SVG -Samsung -Sanyo -Scalado -Sigma -SigmaRaw -Sony -SonyIDC -Stim -SubIFD -System -Theora -Torrent -UserData -VCalendar -VCard -Version0 -Vorbis -WTV -XML -XMP -XMP-DICOM -XMP-Device -XMP-GAudio -XMP-GCamera -XMP-GCreations -XMP-GDepth -XMP-GFocus -XMP-GImage -XMP-GPano -XMP-GSpherical -XMP-LImage -XMP-MP -XMP-MP1 -XMP-PixelLive -XMP-aas -XMP-acdsee -XMP-album -XMP-apple-fi -XMP-aux -XMP-cc -XMP-cell -XMP-crd -XMP-creatorAtom -XMP-crs -XMP-dc -XMP-dex -XMP-digiKam -XMP-drone-dji -XMP-dwc -XMP-exif -XMP-exifEX -XMP-expressionmedia -XMP-extensis -XMP-fpv -XMP-getty -XMP-ics -XMP-iptcCore -XMP-iptcExt -XMP-lr -XMP-mediapro -XMP-microsoft -XMP-mwg-coll -XMP-mwg-kw -XMP-mwg-rs -XMP-pdf -XMP-pdfx -XMP-photomech -XMP-photoshop -XMP-plus -XMP-pmi -XMP-prism -XMP-prl -XMP-prm -XMP-pur -XMP-rdf -XMP-swf -XMP-tiff -XMP-x -XMP-xmp -XMP-xmpBJ -XMP-xmpDM -XMP-xmpMM -XMP-xmpNote -XMP-xmpPLUS -XMP-xmpRights -XMP-xmpTPg -ZIP /path/to/output.jpg
```

However this would throw an confusing error about the output file already existing. I tracked it down to the `-Canon` argument. Not sure why that's causing issues, but when it is removed it works.

So the final command became:

```
exiftool -m -TagsFromFile /path/to/ExifTool.jpg -AC3 -AFCP -AIFF -APE -ASF -AVI1 -Adobe -AdobeCM -AdobeDNG -Apple -Audible -CIFF -CameraIFD -CanonCustom -CanonRaw -CanonVRD -Casio -Composite -DICOM -DJI -DNG -DV -DjVu -DjVu-Meta -Ducky -EPPIM -EXE -EXIF -ExifIFD -ExifTool -FITS -FLAC -FLIR -File -Flash -FlashPix -Font -FotoStation -FujiFilm -FujiIFD -GE -GIF -GIMP -GPS -GeoTiff -GlobParamIFD -GoPro -GraphConv -H264 -HP -HTC -HTML -HTML-dc -HTML-ncc -HTML-office -HTML-prod -HTML-vw96 -HTTP-equiv -ICC-chrm -ICC-clrt -ICC-header -ICC-meas -ICC-meta -ICC-view -ICC_Profile -ID3 -ID3v1 -ID3v1_Enh -ID3v2_2 -ID3v2_3 -ID3v2_4 -IFD0 -IFD1 -IPTC -ISO -ITC -Insta360 -InteropIFD -ItemList -JFIF -JFXX -JPEG -JPEG-HDR -JSON -JVC -Jpeg2000 -KDC_IFD -Keys -Kodak -KodakBordersIFD -KodakEffectsIFD -KodakIFD -KyoceraRaw -LNK -Leaf -LeafSubIFD -Leica -Lytro -M2TS -MAC -MIE-Audio -MIE-Camera -MIE-Canon -MIE-Doc -MIE-Extender -MIE-Flash -MIE-GPS -MIE-Geo -MIE-Image -MIE-Lens -MIE-Main -MIE-MakerNotes -MIE-Meta -MIE-Orient -MIE-Preview -MIE-Thumbnail -MIE-UTM -MIE-Unknown -MIE-Video -MIFF -MNG -MOBI -MOI -MPC -MPEG -MPF0 -MPImage -MS-DOC -MXF -MacOS -MakerNotes -MakerUnknown -Matroska -MediaJukebox -Meta -MetaIFD -Microsoft -Minolta -MinoltaRaw -Motorola -NITF -Nikon -NikonCapture -NikonCustom -NikonScan -Nintendo -Ocad -Ogg -Olympus -OpenEXR -Opus -PDF -PICT -PNG -PNG-pHYs -PSP -Palm -Panasonic -PanasonicRaw -Parrot -Pentax -PhaseOne -PhotoCD -PhotoMechanic -Photoshop -PictureInfo -PostScript -PreviewIFD -PrintIM -ProfileIFD -Qualcomm -QuickTime -RAF -RAF2 -RIFF -RMETA -RSRC -RTF -Radiance -Rawzor -Real -Real-CONT -Real-MDPR -Real-PROP -Real-RA3 -Real-RA4 -Real-RA5 -Real-RJMD -Reconyx -Red -Ricoh -SPIFF -SR2 -SR2DataIFD -SR2SubIFD -SVG -Samsung -Sanyo -Scalado -Sigma -SigmaRaw -Sony -SonyIDC -Stim -SubIFD -System -Theora -Torrent -UserData -VCalendar -VCard -Version0 -Vorbis -WTV -XML -XMP -XMP-DICOM -XMP-Device -XMP-GAudio -XMP-GCamera -XMP-GCreations -XMP-GDepth -XMP-GFocus -XMP-GImage -XMP-GPano -XMP-GSpherical -XMP-LImage -XMP-MP -XMP-MP1 -XMP-PixelLive -XMP-aas -XMP-acdsee -XMP-album -XMP-apple-fi -XMP-aux -XMP-cc -XMP-cell -XMP-crd -XMP-creatorAtom -XMP-crs -XMP-dc -XMP-dex -XMP-digiKam -XMP-drone-dji -XMP-dwc -XMP-exif -XMP-exifEX -XMP-expressionmedia -XMP-extensis -XMP-fpv -XMP-getty -XMP-ics -XMP-iptcCore -XMP-iptcExt -XMP-lr -XMP-mediapro -XMP-microsoft -XMP-mwg-coll -XMP-mwg-kw -XMP-mwg-rs -XMP-pdf -XMP-pdfx -XMP-photomech -XMP-photoshop -XMP-plus -XMP-pmi -XMP-prism -XMP-prl -XMP-prm -XMP-pur -XMP-rdf -XMP-swf -XMP-tiff -XMP-x -XMP-xmp -XMP-xmpBJ -ZIP /path/to/output.jpg
```

This succeeds, but there are lots of groups missing in the resulting file, despite being specified in the command.

I ran the following command to dump both ExifTool.jpg and my target image:

```
exiftool -U -G3:1 -api requestall=3 -api largefilesupport -args /path/to/file.jpg
```

I then used regex to extract the group names for each, sorted them, and deduplicated. The original ExifTool.jpg has the following groups:

```
AVI1
Adobe
AdobeCM
CIFF
CanonVRD
Composite
ExifIFD
ExifTool
File
FlashPix
FotoStation
GraphConv
ICC-header
ICC_Profile
IFD0
IFD1
IPTC
IPTC2
JFIF
JFXX
JPEG-HDR
MIE-Doc
MIE-Main
MPF0
MPImage2
MediaJukebox
MetaIFD
NITF
PhotoMechanic
Photoshop
PictureInfo
PrintIM
Qualcomm
RMETA
SPIFF
Samsung
System
XMP-dc
XMP-photoshop
XMP-rdf
XMP-x
XMP-xmpBJ
XMP-xmpMM
XMP-xmpRights
```

The output file only has these groups:

```
Adobe
CanonVRD
Composite
ExifIFD
ExifTool
File
ICC-header
ICC_Profile
IFD0
IFD1
IPTC
JFIF
System
XMP-dc
XMP-photoshop
XMP-rdf
XMP-x
XMP-xmpBJ
XMP-xmpMM
XMP-xmpRights
```

So it seems like a lot of groups aren't getting block copied for some reason, even when explicitly mentioned.

**Comment by boardhead** [comment]
The -canon causes a problem because Canon is a shortcut tag which includes FileName, so you are trying to rename the file if you include this argument.

- Phil
